icc-x-api/icc-crypto-x-api.ts

```diff
@@ -1243,18 +1243,95 @@ export class IccCryptoXApi {
             }
           )
         : Promise.resolve([])
-      ).then((extractedKeys) =>
-        (hcp as HealthcareParty).parentId
-          ? this.extractKeysFromDelegationsForHcpHierarchy((hcp as HealthcareParty).parentId!, objectId, delegations).then((parentResponse) =>
-              _.assign(parentResponse, {
-                extractedKeys: parentResponse.extractedKeys.concat(extractedKeys),
-              })
-            )
-          : { extractedKeys: extractedKeys, hcpartyId: dataOwnerId }
       )
+        .then(async (extractedKeys) => {
+          if (extractedKeys.length == 0) {
+            const newExtractedKeys = await this._extractDelegationsKeysUsingDataOwnerDelegateAesExchangeKeys(hcp, delegations, objectId)
+            extractedKeys.push(...newExtractedKeys)
+          }
+
+          return extractedKeys
+        })
+        .then((extractedKeys) =>
+          (hcp as HealthcareParty).parentId
+            ? this.extractKeysFromDelegationsForHcpHierarchy((hcp as HealthcareParty).parentId!, objectId, delegations).then((parentResponse) =>
+                _.assign(parentResponse, {
+                  extractedKeys: parentResponse.extractedKeys.concat(extractedKeys),
+                })
+              )
+            : { extractedKeys: extractedKeys, hcpartyId: dataOwnerId }
+        )
     )
   }
 
+  /**
+   * This method covers the use case when a DataOwner found back access to its data, and needs to use the delegations of its delegates instead
+   * of its own ones
+   *
+   * @param dataOwner The current data owner from whom we want to decrypt the aesExchangeKeys
+   * @param delegations The object delegations to decrypt
+   * @param objectId The object to decrypt id
+   * @private
+   */
+  private async _extractDelegationsKeysUsingDataOwnerDelegateAesExchangeKeys(
+    dataOwner: Patient | Device | HealthcareParty,
+    delegations: { [p: string]: Array<Delegation> },
+    objectId: string
+  ) {
+    // Find other keys through aesExchangeKeys
+    const dataOwnerPubKeys = await this.getPublicKeys()
+    const keysToDecrypt = fold(
```

Comment:
I imagine you use fold instead of reduce because it is some kotlin code that has been ported ?